### PR TITLE
Clean up QueryParseContext and don't hold it inside QueryRewrite/ShardContext

### PR DESCRIPTION
**Pull request by cbuescher**
This change cleans up a few things in QueryParseContext and QueryShardContext that make it hard to reason about the state of these context objects and are thus error prone and should be simplified.

Currently the parser that used in QueryParseContext can be set and reset any time from the outside, which makes reasoning about it hard.
This change makes the parser a mandatory constructor argument removes ability to later set a differen ParseFieldMatcher. If none is provided at construction time, the one set inside the parser is used. If a ParseFieldMatcher is specified at construction time, it is implicitely set on the parser that is beeing used. The ParseFieldMatcher is only kept inside the parser.

Also currently the QueryShardContext historically holds an inner QueryParseContext (in the super class QueryRewriteContext), that is mainly used to hold the parser and parseFieldMatcher. For that reason, the parser can also be reset, which leads to the same problems as above. This change removes the QueryParseContext from QueryRewriteContext and removes the ability to reset or retrieve it from the QueryShardContext. Instead, `QueryRewriteContext#newParseContext(parser)` can be used to create new parse contexts with the given parser from a shard context when needed.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java [hunk 1]
```diff
@@ -42,7 +43,6 @@ public QueryRewriteContext(IndexSettings indexSettings, MapperService mapperServ
         this.scriptService = scriptService;
         this.indexSettings = indexSettings;
         this.indicesQueriesRegistry = indicesQueriesRegistry;
-        this.parseContext = new QueryParseContext(indicesQueriesRegistry);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java]: so we had this instance member but it was unused? :)

Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java]: oh I see better now that I opened up the code....what a mess :)

File: core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java [hunk 2]
```diff
@@ -26,42 +26,33 @@
...
 
-    private XContentParser parser;
-    private ParseFieldMatcher parseFieldMatcher = ParseFieldMatcher.EMPTY;
+    private final XContentParser parser;
 
     private IndicesQueriesRegistry indicesQueriesRegistry;
 
-    public QueryParseContext(IndicesQueriesRegistry registry) {
-        this.indicesQueriesRegistry = registry;
+    public QueryParseContext(IndicesQueriesRegistry registry, XContentParser parser, ParseFieldMatcher parseFieldMatcher) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: does it still make sense to use a parse field matcher that is not the one set to the parser at this point ? maybe too big of a change for now?

File: core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java [hunk 3]
```diff
@@ -26,42 +26,33 @@
...
+        this.indicesQueriesRegistry = Objects.requireNonNull(registry, "indices queries registry cannot be null");
+        this.parser = Objects.requireNonNull(parser, "parser cannot be null");
+        this.parser.setParseFieldMatcher(Objects.requireNonNull(parseFieldMatcher, "parse field matcher cannot be null"));
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: refers to the other comment above, doesn't feel right to mutate one of the arguments here based on the second argument. Can we require that the parse field matcher is properly set to the parser all the time instead?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: fixing this would allow us to get rid of setParseFieldMatcher in XContentParser and make the matcher final in AbstractXContentParser. I think this should be the direction moving forward, what do you think?

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java [hunk 4]
```diff
@@ -114,28 +113,21 @@ public QueryShardContext(QueryShardContext source) {
...
-    public void parseFieldMatcher(ParseFieldMatcher parseFieldMatcher) {
-        this.parseContext.parseFieldMatcher(parseFieldMatcher);
-    }
-
+    /**
+     * @return the parseFieldMatcher configured in the index settings
+     */
     public ParseFieldMatcher parseFieldMatcher() {
-        return parseContext.parseFieldMatcher();
+        return indexSettings.getParseFieldMatcher();
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: I hadn't noticed this before, but isn't there a circular dependency between IndexSettings and ParseFieldMatcher. I think I would remove the instance within IndexSettings if possible. Doesn't seem hard.

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java [hunk 5]
```diff
@@ -114,28 +113,21 @@ public QueryShardContext(QueryShardContext source) {
...
     }
 
     public void reset() {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: do you think it would be possible to get rid of this reset too and make sure that each operation that needs a reset actually uses a new instance of the context? Seems much cleaner.

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java [hunk 6]
```diff
@@ -343,38 +331,35 @@ public ParsedQuery parse(XContentParser parser) {
      */
     @Nullable
     public ParsedQuery parseInnerFilter(XContentParser parser) throws IOException {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: Also, do you think it's feasible now to take all the parsing stuff out of this context? maybe the few components left that need this should rather take in as argument both shard context and parse context and use them properly instead of relying on shard context to parse through parse context?

**Comment by javanna** [comment]
I like these changes a lot! I left some comments, mainly ideas to explore I think, but they don't need to be all implemented as part of this PR. Up to you which ones can be done here and which ones are bigger and deserve their own followup PR.

One more thing, can we make the ParseFieldMatcher constructor that takes a boolean argument private?

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java (hunk 4 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: So you suggest removing the parseFieldMatcher from IndexSettings and and instead hold an instance insice QueryShardContext instead? As far as I can see we need IndexSettings in QueryShardContext for the super ctor anyway, so my idea was to have it only in one place. Keeping the "real" instance in IndexSettings has the advantage that it only has to be creates once, while when moving it to QueryShardContext, we will have to create it in the ctor every time, but maybe I missunderstand what you mean.

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java (hunk 5 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: I'll look into it, but maybe to keep this change from growing too large I would tackle that removal as a follow up.

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java (hunk 6 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: I agree, unfortunately these are used all over the place. I will have another look but again maybe that is better suited for a follow up PR.

File: core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java (hunk 2 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: I added this mainly because in tests we sometimes use ParseFields.EMPTY, sometimes ParseFields.STRICT. I think it would generally be much better to only keep the ParseFieldMatcher in one place, e.g. the parser itself, so I will look if I can get rid of this.

File: core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java (hunk 3 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: Again I agree, just couldn't decide where the best place for the one "authoritative" source of the parse field matcher that is used should be. I'll see if we can have it only in the parser.

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java (hunk 4 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: forget about it, ParseFieldMatcher depends on Settings not IndexSettings, no circular dep, no problem! Leave it as is, your explanation makes perfect sense.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: I wish we were creating less instances of ParseFieldMatcher actually, but that's another problem :)

File: core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java (hunk 3 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: So just from taking a brief look at this, this going to be a big PITA: making the matcher final in AbstractXContentParser would mean we need it as ctor argument, which means we need the index setting everwhere we create a parser, which is basically in a gazillion of places. Actually the parseFieldMatcher inside AbstractXContentParser was added just recently, I think mainly to be available in ObjectParser and RestReindexAction, where we up to now dont have a ParseContext. The problem is that currently the ParseFieldMatcher in the context and the one used in the parser might diverge.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: ok, I haven't followed all recent changes, I just realized that there are ParseFieldMatcher instances in different places. I also wonder now if it should be in the parser itself, like should it be in the parser or in the parse context? We should pick one, it cannot be both ideally :)

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: As I said, ideally I'd say its the parser, but thats a huge change. Also one might argue that the parsers job is to keep state for parsing, but the matcher is something the client (e.g. the builders fromXContent) use, so there is just the question of how to transport it there, so that would be the context. I'm just looking into how difficult it would be to remove the parseField from the parser again, this duplication is a source of errors.

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java (hunk 4 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: ++ to that. In the tests its good though that some methods take the matcher as an argument, so we can test thing in strict and non-strict mode without having pass IndexSettings around.

File: core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java (hunk 2 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: Comming back to this: the problem with the parseFieldMatcher stored inside AbstractXContentParser is that (currently on maste) we don't set it anywhere except in ObjectParserTests and when the parser is reset() on the QueryParseContext (where its broken btw.). I agree that it would be best if there was only one ParseFieldMatcher (either the one set on the parser, or the one in the QueryParseContext). 

Having PFM only in the parser is hard because:
* we'd ideally make parseFieldMatcher a ctor arg for all XContentParsers. This however is a huge change (tons of ways to create XContentParser, all would need to be changes), I'm not even sure if this would work because we would need the IndexSettings (where the flag for strict parsing comes from) is held
* if we keep the current get/setParseFieldMatcher on XContentParser it is hard to check that we alway set it correctly

On the other hand, keeping PFM only in the QueryShardContext requires us removing it from ObjectParser. I ran into many problems trying to do this (to me it seems the setters, getters where just introduced in 5f258b to make PFM play nicely with ObjectParser). 

So for now I'm a little bit stuck here. I'm leaning toward trying to removing the PFM from the parser and only store it in the QueryParseContext, but I'm struggling to remove it there because ObjectParser depends and so far I didn't find a way to get around the getter/setter on the parser.

File: core/src/main/java/org/elasticsearch/index/percolator/PercolatorFieldMapper.java [hunk 7]
```diff
@@ -190,22 +188,18 @@ static Query toQuery(QueryShardContext context, boolean mapUnmappedFieldsAsStrin
         // as an analyzed string.
         context.setAllowUnmappedFields(false);
         context.setMapUnmappedFieldAsString(mapUnmappedFieldsAsString);
-        context.parseFieldMatcher(context.getIndexSettings().getParseFieldMatcher());
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/percolator/PercolatorFieldMapper.java]: The shard context parseFieldMatcher IMHO should always be backed by `indexSettings.getParseFieldMatcher()`

**Comment by cbuescher** [comment]
@javanna I rebased this PR on current master and added a commit that now keeps the parseFieldMatcher only in the parser and not storing it in the QueryParseContext. The `parseFieldMatch()` getter of the context just forwards to the parser. This however makes is necessary that we alway set the parseFieldMatcher in the parsers instead of now where we set it on the context. In the future we should try to make ParseFieldMatcher immutable on parsers and make it a mandatory constructor argument of AbstractXContentParser. But this is a big change and out of the scope of this PR.

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java [hunk 8]
```diff
@@ -343,38 +331,38 @@ public ParsedQuery parse(XContentParser parser) {
...
         } finally {
-            reset(null);
+            reset();
         }
     }
 
-    public Query parseInnerQuery() throws IOException {
-        return toQuery(this.parseContext().parseInnerQueryBuilder(), this);
+    public Query parseInnerQuery(XContentParser parser) throws IOException {
+        parser.setParseFieldMatcher(this.parseFieldMatcher());
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: it is odd that we do the opposite here, as the context holds the parse field matcher and sets it to the parser. But I do see why. The parse field matcher is used for more than just parsing xcontent. I think it's ok then.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java [hunk 9]
```diff
@@ -162,8 +162,8 @@ protected boolean doEquals(WrapperQueryBuilder other) {
     @Override
     protected QueryBuilder<?> doRewrite(QueryRewriteContext context) throws IOException {
         try (XContentParser qSourceParser = XContentFactory.xContent(source).createParser(source)) {
-            QueryParseContext parseContext = context.newParseContext();
-            parseContext.reset(qSourceParser);
+            qSourceParser.setParseFieldMatcher(context.indexSettings.getParseFieldMatcher());
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: do we want to expose getParseFieldMatcher in QueryRewriteContext that takes it from indexSettings? then we would not need it in QueryShardContext as it's a subclass?

File: core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java [hunk 10]
```diff
@@ -37,10 +37,9 @@
 
     private final IndicesQueriesRegistry indicesQueriesRegistry;
 
-    public QueryParseContext(IndicesQueriesRegistry registry, XContentParser parser, ParseFieldMatcher parseFieldMatcher) {
+    public QueryParseContext(IndicesQueriesRegistry registry, XContentParser parser) {
         this.indicesQueriesRegistry = Objects.requireNonNull(registry, "indices queries registry cannot be null");
         this.parser = Objects.requireNonNull(parser, "parser cannot be null");
-        this.parser.setParseFieldMatcher(Objects.requireNonNull(parseFieldMatcher, "parse field matcher cannot be null"));
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: By moving this out we need to make sure the setter is called in many places where we previously implicitly set it via `context.parseFieldMatcher(pfm)`. I only checked the places where previously this method was calles somehow. 
I'm on the fence if maybe we should instead keep it here, that way at least we know there is one location where the parseFieldMatcher is set.

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregatorParsers.java [hunk 11]
```diff
@@ -131,6 +131,7 @@ public AggregatorParsers(ParseFieldRegistry<Aggregator.Parser> aggregationParser
                     if (token == XContentParser.Token.VALUE_STRING || token == XContentParser.Token.VALUE_EMBEDDED_OBJECT) {
                         byte[] source = parser.binaryValue();
                         binaryParser = XContentFactory.xContent(source).createParser(source);
+                        binaryParser.setParseFieldMatcher(parseContext.parseFieldMatcher());
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregatorParsers.java]: seeing this I am wondering if this getter makes it confusing, shall we consider dropping it in favour of doing parseContext.parser().parseFieldMatcher() ? Or maybe it's not a big deal?

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java (hunk 9 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: good point, I think thats a good idea. Only one is called parseFieldMatcher, the other getParseFieldMatcher, I guess we are leaning to getParseFieldMatcher then.

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregatorParsers.java (hunk 11 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregatorParsers.java]: I think its okay, maybe we can add a comment to the getter that makes clear its a shortcut to the matcher in the parser.

File: core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java [hunk 12]
```diff
@@ -713,19 +713,14 @@ protected QB copyQuery(QB query) throws IOException {
      * @return a new {@link QueryShardContext} based on the base test index and queryParserService
      */
     protected static QueryShardContext createShardContext() {
-        QueryShardContext queryCreationContext = new QueryShardContext(queryShardContext);
-        queryCreationContext.reset();
-        queryCreationContext.parseFieldMatcher(ParseFieldMatcher.STRICT);
-        return queryCreationContext;
+        return new QueryShardContext(queryShardContext);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: are we losing the STRICT bit here? it's important that we use STRICT here, so we make sure that we never output deprecated stuff ourselves. and we test deprecations separately.

File: core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java [hunk 13]
```diff
@@ -713,19 +713,14 @@ protected QB copyQuery(QB query) throws IOException {
...
 
     /**
      * @return a new {@link QueryParseContext} based on the base test index and queryParserService
      */
-    protected static QueryParseContext createParseContext() {
-        QueryParseContext queryParseContext = new QueryParseContext(indicesQueriesRegistry);
-        queryParseContext.reset(null);
-        queryParseContext.parseFieldMatcher(ParseFieldMatcher.STRICT);
+    protected static QueryParseContext createParseContext(XContentParser parser) {
+        QueryParseContext queryParseContext = new QueryParseContext(indicesQueriesRegistry, parser);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: same here, might be that we are good, but let's make sure we don't lose the STRICT one

**Comment by cbuescher** [comment]
@javanna so I think after opening #17756 (and after that gets some traction) I'm going to rework this and remove the last commit.

File: core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java (hunk 12 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: This is a shard context, so these parse field matcher should always be backed by indexSettings. If we want to default to STRICT here we should change that in the setup of the test I think.

**Comment by cbuescher** [comment]
@javanna after merging #17756 I updated this and removed some more usages of the STRICT constant from core. I think this is much cleaner now.

File: core/src/main/java/org/elasticsearch/action/percolate/TransportPercolateAction.java [hunk 14]
```diff
@@ -196,9 +196,8 @@ public static SearchRequest createSearchRequest(PercolateRequest percolateReques
 
         PercolatorQueryBuilder percolatorQueryBuilder = new PercolatorQueryBuilder(percolateRequest.documentType(), documentSource);
         if (querySource != null) {
-            QueryParseContext queryParseContext = new QueryParseContext(queryRegistry);
-            queryParseContext.reset(XContentHelper.createParser(querySource));
-            queryParseContext.parseFieldMatcher(parseFieldMatcher);
+            QueryParseContext queryParseContext = new QueryParseContext(queryRegistry, XContentHelper.createParser(querySource),
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/percolate/TransportPercolateAction.java]: just noticed that below we have try finally around the parser creation but here we don't have it. Should we have that here too? who closes the parser after all?

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java [hunk 15]
```diff
@@ -115,18 +114,12 @@ public QueryShardContext(QueryShardContext source) {
 
     public void reset() {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: please please can we look into removing this as well later? :)

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java (hunk 6 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: ++

File: core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java (hunk 12 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: this is fixed by adding the strict setting above in the init method right?

File: core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java [hunk 16]
```diff
@@ -158,11 +157,6 @@
     private static int queryNameId = 0;
     private static SearchModule searchModule;
 
-
-    protected static QueryShardContext queryShardContext() {
-        return queryShardContext;
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: I am now wondering if we still need this queryShardContext instance member if we (almost) always create a new one.

File: core/src/test/java/org/elasticsearch/search/searchafter/SearchAfterBuilderTests.java [hunk 17]
```diff
@@ -224,7 +221,7 @@ public void testFromXContent() throws Exception {
             searchAfterBuilder.innerToXContent(builder);
             builder.endObject();
             XContentParser parser = XContentHelper.createParser(builder.bytes());
-            context.reset(parser);
+            QueryParseContext context = new QueryParseContext(indicesQueriesRegistry, parser, ParseFieldMatcher.STRICT);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/searchafter/SearchAfterBuilderTests.java]: thanks, it is better to use strict than the default one here!

File: core/src/test/java/org/elasticsearch/search/sort/SortParserTests.java [hunk 18]
```diff
@@ -1,165 +0,0 @@
...
-import org.elasticsearch.common.xcontent.XContentHelper;
-import org.elasticsearch.common.xcontent.XContentParser;
-import org.elasticsearch.index.IndexService;
-import org.elasticsearch.index.query.QueryParseContext;
-import org.elasticsearch.test.ESSingleNodeTestCase;
-import org.elasticsearch.test.TestSearchContext;
-
-import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
-
-public class SortParserTests extends ESSingleNodeTestCase {
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/sort/SortParserTests.java]: oh nice we had tests for a class that didn't exist anymore :) thanks for fixing

File: core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java (hunk 15 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java]: I see that I had already asked this before, ignore me, you already said we'll do it in another PR :)

**Comment by javanna** [comment]
this looks much cleaner indeed! left a few minors but LGTM!

File: core/src/main/java/org/elasticsearch/action/percolate/TransportPercolateAction.java (hunk 14 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/percolate/TransportPercolateAction.java]: Sure, I see parse creation a lot as try-with-resouces without a finally block (unless we need to reset something which here I think we don't), so I will do that.

File: core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java (hunk 16 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: Its good to have it as a template for the createShardContext() method, we also need the instance member because we use it in lambdas for the mock mapping service etc...

Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: it gets a bit weird when we use it to set stuff to the SearchContext although we create another context for those operations. That seems confusing... plus if it was not for templating, it would be only needed in init() which could have it a local variable. Maybe rename it at least so it is clearer why it's there? and fix some of its usages?

Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: okay, so I understand you correctly: should I revert 4e77adf38 instead and we rather use the instance member as much as possible and try to remove the `createShardContext()` that makes a copy instead?

Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: no  I like creating a new instance, seems to reflect what we do everywhere at this point. But we should clean things up a little bit more, to make sure that the set search context is consistent at least.

File: core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java (hunk 12 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: yes

File: core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java (hunk 16 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: Okay, I think I get it. I went ahead and removed the queryShardContext instance entirely. I had to introduce a few more static fields for all the services we need in init(), but I think it is better this way. Also pulled up some mocking on to of the SearchContext that was done in the Parent/Child/Nested tests but that can be used everywhere I think.

**Comment by cbuescher** [comment]
@javanna thanks for the last round of comments, I adressed them in the last two commits.

File: core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java (hunk 16 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: thanks a lot, looks great!

**Comment by javanna** [comment]
LGTM
